### PR TITLE
KEYCLOAK-5098 - adapter for spring boot 2

### PR DESCRIPTION
**Pull request by ryandawsonuk**
New PR to replace https://github.com/keycloak/keycloak/pull/4279 as rebase didn't go well.

As per that PR, Spring boot 2.0.0 is in M2 at the time of writing but this is tested and working with Jetty and Undertow. There is a problem with boot2 specific to Tomcat.

The adapter is a port of the existing boot 1 adapter but adjusted for the changes in boot 2. An spring-boot-adapter-core library has been created so that all of the common code is reused across the two adapter versions.

### CODE REVIEW

**Comment by vitornp** [comment]
Forecast for next release with PR?

**Comment by ryandawsonuk** [comment]
This PR includes the changes requested in the referenced one so it's ready from my perspective.

**Comment by sebastienblanc** [comment]
I will try to review this early this week, @vitornp hopefully it will be included in the next release (and later when SB 2.0 is out I will send a PR to start.spring.io to update the starter)

File: adapters/oidc/spring-boot2/src/main/java/org/keycloak/adapters/springboot/KeycloakSpringBootProperties.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+import org.keycloak.representations.adapters.config.AdapterConfig;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@ConfigurationProperties(prefix = "keycloak", ignoreUnknownFields = false)
+public class KeycloakSpringBootProperties extends AdapterConfig {
```
Comment (vitornp) [inline adapters/oidc/spring-boot2/src/main/java/org/keycloak/adapters/springboot/KeycloakSpringBootProperties.java]: Same file in keycloak-spring-boot-adapter-core

**Comment by vitornp** [comment]
Try version with PR when start show error:
```
***************************
APPLICATION FAILED TO START
***************************

Description:

Binding to target [Bindable@124eb83d type = java.util.Map<?, ?>, value = 'provided', annotations = array<Annotation>[@com.fasterxml.jackson.annotation.JsonIgnore(value=true)]] failed:

    Reason: Failed to bind properties under 'keycloak.config' to java.util.Map<?, ?>

Action:

Update your application's configuration
```

**Comment by jmcshane** [comment]
I'm not a fan of the adapter module names (springboot vs. spring-boot-2). Could we retain the original folder name of spring-boot?

**Comment by ryandawsonuk** [comment]
Apologies, I had a commit that I hadn't pushed.

@vitornp That class is no longer duplicated - see https://github.com/ryandawsonuk/keycloak/tree/master/adapters/oidc/spring-boot2/src/main/java/org/keycloak/adapters/springboot
and https://github.com/ryandawsonuk/keycloak/tree/master/adapters/oidc/spring-boot2/src/main/java/org/keycloak/adapters/springboot

Not yet looked at the application failing to start. I did check that prior to my last merge and it was working then. I suspect it's not related to my changes as my understanding is that the adapter is just for clients.

@jmcshane The only instances of 'springboot' without the hyphen now are in the package names, which was the case before these changes.

**Comment by ryandawsonuk** [comment]
Have now tried starting the built distribution from standalone.sh as per the README (is that what you meant?) and I'm not seeing that 'failed to start' message - am able to start and reach the admin console.

**Comment by ryandawsonuk** [comment]
@vitornp , @jmcshane , @sebastienblanc do let us know if there's anything more you're looking for on this. We're already using this adapter with v7 of Activiti BPM (https://github.com/Activiti/activiti-examples/tree/master/spring-boot-sample-activiti) and would like to see it in keycloak rather than building the adapter ourselves. I see it's scheduled for 3.3.0.CR1 so let us know if that changes.

**Comment by sebastienblanc** [comment]
@ryandawsonuk Have you made sure it works still correctly with Spring Boot 1.x ? If yes, I'm okay with the PR and we email the core commiter to merge.

**Comment by ryandawsonuk** [comment]
Yes I've also checked that I can use the refactored version of the 1.x adapter to secure a spring boot as per https://developers.redhat.com/blog/2017/05/25/easily-secure-your-spring-boot-applications-with-keycloak/

**Comment by Salaboy** [comment]
@sebastienblanc @ryandawsonuk I have a question for the experts, which is extremely related with this PR. Is the adapter forwarding the original request headers for example: X-FORWARDED-*?
Can this be configured somewhere in this adapters?

**Comment by stianst** [comment]
Thanks for this PR. We don't have the capacity to get Spring Boot 2 support reviewed and merged until we start on Keycloak 4.x.

**Comment by Salaboy** [comment]
@stianst thanks for the feedback. Can we have some kind of estimate on when you guys plan to get started with keycloak 4.x ? Spring boot 2 is going to be released Nov/Dec so it will be great to have this merged at some point after it is released. For now, we are building this integration internally as part of our project [1] so we can use it. 

[1]: activiti.org

**Comment by stianst** [comment]
Probably Nov/Dec

**Comment by sebastienblanc** [comment]
Hi all ! I finally have some time this sprint to work on this. 
@ryandawsonuk I rebased your branch and did a new PR for testing purposed there : https://github.com/keycloak/keycloak/pull/4976 

Looks pretty good, my test app with Tomcat embedded is working. Should we continue there ?

**Comment by Salaboy** [comment]
@sebastienblanc can you check with jetty as well? that's the default for spring boot as far as I can remember. I want to make sure that people using jetty can get it working as well.

**Comment by sebastienblanc** [comment]
@Salaboy The default is Tomcat , brought by `spring-boot-starter-web` but for sure I will test with Jetty as well.

**Comment by Salaboy** [comment]
@sebastienblanc you are right, I got it backwards, but it need to work with both

**Comment by iamsharife** [comment]
Is this going to be merged? or if not - what do i need to do to get Keycloak to work in Spring Boot 2?

**Comment by logemann** [comment]
Any status on that? Spring Boot 2 being final for some days now.... If it will take some time i need to downgrade my app to Spring Boot 1

**Comment by sebastienblanc** [comment]
It continues here https://github.com/keycloak/keycloak/pull/4976

**Pull request closed without merging**